New JSON Endpoint & refactor to use Map[String,String] instead of LogEvent

CattMarlin opened the pull request:
* LogCollectorRoutes
    * New endpoint that parses JSON Array into Array[Map[String,String]], then sends each Map to the stream
* LogCollectorRoutesTest:
    * test that takes a JSON and runs it through the new route
* Added Util Object:
    * A function logEventToFlattenedMap to convert & flatten LogEvent to Map[String, String]
* Added UtilTest:
    * UtilTest tests the function in Util logEventToFlattenedMap on a LogEvent
    * UtilTest tests the function in Util logEventToFlattenedMap on a LogEvent with properties
* SolrCloudStreams:
    * now takes Map[String,String]
* SolrCloudStreamsTest:
    * test that takes Map[String,String] doc and writes to the collection, asserting that the doc count is greater than 0
* DocumentConversion:
    * Added a function to convert Map[String,String] to SolrDocument
* AlertEngineImplTest & SolrServiceTest:
    * Update to not use DocumentConversion.toSolrDocument, use DocumentConversion.mapToSolrDocument instead
* PulseKafkaConsumer:
    * Incorporated logEventToFlattenedMap
* Schema.xml, secure Schema.xml, test Schema.xml:
    * Changed all required fields except id to required=false.

afoerster commented on alert-engine/src/test/scala/io/phdata/pulse/alertengine/AlertEngineImplTest.scala:
```diff
@@ -64,16 +65,25 @@ class AlertEngineImplTest
+        "message"   -> "message2",
+        "throwable" -> "Exception in thread main"))
+
+    println(documentError.toString)
```
Nit: remove println (unless there's a good reason for it)

afoerster commented on alert-engine/src/test/scala/io/phdata/pulse/alertengine/AlertEngineImplTest.scala:
```diff
@@ -85,9 +95,9 @@ class AlertEngineImplTest
-                                                  retryInterval = 1,
-                                                  resultThreshold = Some(1),
-                                                  alertProfiles = List("test@phdata.io"))
+      retryInterval = 1,
```
Did you run `sbt scalafmt` before checking this in? These should have already been formatted. I'm wondering if this change was caused by the Intellij formatting.

CattMarlin replied:
Weird.  I didn't touch this line.  I ran `make package` before commit.  I just ran `sbt scalafmt` and it stayed the same.

afoerster commented on log-collector/src/main/scala/io/phdata/pulse/logcollector/LogCollectorRoutes.scala:
```diff
@@ -79,10 +81,27 @@ class LogCollectorRoutes(solrService: SolrService) extends JsonSupport with Lazy
+     */
+    post {
+      // create a streaming Source from the incoming json string
+      entity(as[String]) { logEvent =>
```
Can you change this to accept an array of events by default? That will always be more efficient. We need to deprecate the endpoint that accepts a single log event.

CattMarlin replied:
Would this not cause a problem when someone sends something that doesn't conform to LogEvent like `{my_metric: 20.2, timestamp: 2019-03-06T01:14:11}`.  Or are you saying it should take `Array[String]` ?

afoerster replied:
A user should be able to send a JsonArray[JsonObject], you can still do your casting to `as[String]`, the change is to parse it into an array of objects and `foreach` them onto the stream

afoerster commented on log-collector/src/main/scala/io/phdata/pulse/logcollector/PulseKafkaConsumer.scala:
```diff
@@ -53,11 +53,12 @@ class PulseKafkaConsumer(solrService: SolrService) extends JsonSupport with Lazy
-            .parseJson
-            .convertTo[LogEvent]
-          solrInputStream ! (logEvent.application.get, logEvent)
+          val logEventMap = Util.logEventToFlattenedMap(
```
There might be a little too much nesting here, maybe pull out the `record.value...` into its own value?

afoerster commented on log-collector/src/main/scala/io/phdata/pulse/logcollector/Util.scala:
```diff
@@ -0,0 +1,34 @@
+   * @param logEvent LogEvent to convert
+   * @return The event, with properties flattened, as a [[Map[String,String]]
+   */
+  def logEventToFlattenedMap(logEvent: LogEvent): Map[String, String] = {
```
Is there a test for this?

CattMarlin replied:
Will add a test

afoerster commented on log-collector/src/main/scala/io/phdata/pulse/logcollector/Util.scala:
```diff
@@ -0,0 +1,34 @@
+   * @return The event, with properties flattened, as a [[Map[String,String]]
+   */
+  def logEventToFlattenedMap(logEvent: LogEvent): Map[String, String] = {
+    val values = logEvent.productIterator
```
Since there are only 7 or 8 fields in a LogEvent, it might be easier and cleaner to just manually create a map from them.

This is a generic case class to Map function, but we only have a single case class, and we know all of the fields, so it seems a little complex for what we are trying to do

CattMarlin replied:
Sounds good.  I wanted to avoid hardcoding information about the case class.  Will rewrite.

afoerster replied:
Understand that. In this case hardcoding is fine. It will be obvious to a reader.

afoerster commented on log-collector/src/main/scala/io/phdata/pulse/logcollector/Util.scala:
```diff
@@ -0,0 +1,34 @@
+object Util {
+
+  /**
+   * Convert a [[LogEvent]] to a flattened [[Map[String,String]]
```
You can be specific in this comment and tell you are flattening out the `properties` field only, without that knowledge the user doesn't know what 'flattening' means.

afoerster commented on log-collector/src/main/scala/io/phdata/pulse/logcollector/Util.scala:
```diff
@@ -0,0 +1,34 @@
+    logEventmap
+      .flatten {
+        case ((key, map: Option[Map[String, String]]))
+            if map.getOrElse(None).isInstanceOf[Map[String, String]] =>
```
I don't think this code should necessarily stay for the reasons above, but all the casting shouldn't be necessary.
Since option has a map type, it might simplify to something more like

```
case ((_, properties: Some[Map[String, String]])) => properties.get
```

`isInstanceOf` is almost always a smell, especially since you can pattern match on types.

afoerster commented on common/src/test/scala/io/phdata/pulse/common/SolrServiceTest.scala:
```diff
@@ -132,25 +131,25 @@ class SolrServiceTest extends FunSuite with BaseSolrCloudTest {
+        "message"    -> "message",
+        "threadName" -> "thread oxb",
+        "throwable"  -> "Exception in thread main"
+      ).asInstanceOf[Map[String, String]])
```
Why the cast? I don't think it's necessary

CattMarlin replied:
Without casting `mapToSolrDocument` complains about the `None`

afoerster replied:
Can you share that error? Solr should be creating its own ids, and I believe that endpoint takes a `SolrDocument` type, so this is odd.

afoerster commented on common/src/test/scala/io/phdata/pulse/common/SolrServiceTest.scala:
```diff
@@ -167,17 +166,17 @@ class SolrServiceTest extends FunSuite with BaseSolrCloudTest {
+        "threadName" -> "thread oxb",
+        "throwable"  -> "Exception in thread main",
+        "hostname"   -> "host1.com"
+      ).asInstanceOf[Map[String, String]])
```
Cast?

CattMarlin replied:
Without casting `mapToSolrDocument` complains about the `id=None`.

CattMarlin replied:
Think you meant to respond to this with:

 "Can you share that error? Solr should be creating its own ids, and I believe that endpoint takes a SolrDocument type, so this is odd."

The error is with `mapToSolrDocument` and it is a `Type Mismatch: Map[String, Serializable] does not conform to type Map[String, String]`.

I will just remove the `"id" -> None`.  I think the way the code works, a `None` would not make it to `mapToSolrDocument` and that was an oversight by me.

afoerster commented on log-collector/src/main/scala/io/phdata/pulse/logcollector/Util.scala:
```diff
@@ -0,0 +1,34 @@
+      }
+      .toMap
+      .filter(kv => !kv._2.equals(None))
+      .asInstanceOf[Map[String, String]]
```
Why is this cast necessary?

CattMarlin replied:
Going to rewrite per you other comment about LogEvent's fields always being known.

CattMarlin replied:
The new "hardcoded" version still does this.  Lets discuss further once I push the changes and you get to take a look. I think I see a way around it but it would be ugly.

afoerster commented on log-collector/src/main/scala/io/phdata/pulse/logcollector/Util.scala:
```diff
@@ -0,0 +1,34 @@
+        case ((key, value)) => Map(key -> value)
+      }
+      .toMap
+      .filter(kv => !kv._2.equals(None))
```
Is it possible to not add these in the first place?

CattMarlin replied:
Going to rewrite per you other comment about LogEvent's fields always being known.

CattMarlin replied:
New "hardcoded" function still does this.  Lets discuss further once I push the changes and you get to take a look.

afoerster replied:
Can the above expressions be `getOrElse("")` instead of `getOrElse(None)`? Then it's an empty string and no check for `None` is needed

afoerster commented on cloudera-integration/parcel/conf/solr-configs/pulseconfig-securev1/schema.xml:
```diff
@@ -19,20 +19,20 @@
 
     <fields>
-        <field name="id" type="string" indexed="true" stored="true" required="true"
+        <field name="id" type="string" indexed="true" stored="true" required="false"
```
I _think_ we need to keep `id` required (but only `id`), because it's the unique key. The key is auto generated. I'm looking here: https://wiki.apache.org/solr/UniqueKey. But since it's auto generated, we can just make it required and not worry about it.

afoerster commented on cloudera-integration/parcel/conf/solr-configs/pulseconfigv2/schema.xml:
```diff
@@ -19,20 +19,20 @@
 
     <fields>
-        <field name="id" type="string" indexed="true" stored="true" required="true"
+        <field name="id" type="string" indexed="true" stored="true" required="false"
```
id should be required, same as above

afoerster commented on log-collector/src/main/scala/io/phdata/pulse/logcollector/LogCollectorRoutes.scala:
```diff
@@ -79,10 +81,30 @@ class LogCollectorRoutes(solrService: SolrService) extends JsonSupport with Lazy
+      entity(as[String]) { logEvent =>
+        logger.trace("received message")
+
+        val parsedJsonArray = logEvent.parseJson.convertTo[Array[Map[String, String]]]
```
👍

afoerster commented on log-collector/src/main/scala/io/phdata/pulse/logcollector/Util.scala:
```diff
@@ -0,0 +1,33 @@
+      .filter(kv => !kv._2.equals(None))
+      .asInstanceOf[Map[String, String]]
+
+    if (logEvent.properties.getOrElse(None).isInstanceOf[Map[String, String]])
```
Case statements are the preferred way to do this:
```
logEvent.properties match {
case Some(properties) => stringMap ++ properties
case None => stringMap
}
```

afoerster commented on log-collector/src/test/scala/io/phdata/pulse/logcollector/UtilTest.scala:
```diff
@@ -0,0 +1,25 @@
+    val parsedLogEventMap = Util.logEventToFlattenedMap(document1)
+
+    assert(parsedLogEventMap.isInstanceOf[Map[String, String]])
+
```
Another test that tests the `properties` flattening would be good here.

CattMarlin commented:
I made the schema config changes, added the send Util test, and reworked `logEventToFlattenedMap` per your comments.  Comment at top reflects all changes to date.

afoerster commented on log-collector/src/main/scala/io/phdata/pulse/logcollector/Util.scala:
```diff
@@ -0,0 +1,31 @@
+      "application" -> logEvent.application.getOrElse("")
+    )
+
+    val stringMap = rawMap
```
This looks like it isn't necessary, they're both values so you can just use rawMap below
